docs/proposals/component-registration.md

```diff
@@ -0,0 +1,378 @@
+
+## Terminology
+
+To clarify, a "component" in this context is a core part of the Kubernetes cluster, required or optional, that knows about Kubernetes and exposes a method for probing the liveness and readiness of each instance.
+
+The set of available/deployed/registered component types and/or instances may change over time.
+
+Currently there are three types of components in a vanilla Kubernetes deployment, each with a single instance: apiserver, controller-manager, and scheduler.
+
+Etcd should not be considered a component, because it is a dependency of the apiserver, does not know about Kubernetes, and thus cannot register itself or be expected to have both liveness and readiness probe endpoints. However, the current `/componentstatuses` endpoint includes etcd, and must continue to do so for reverse compatibility.
+
+
+## Use Cases
+
+This proposal primarily covers features that can be achieved internally by Kubernetes itself, but it's clear that there are also some desirable features that must be provided by external deployment, management, or monitoring systems. These external systems, however, require some features within Kubernetes to enable them. So while not all of these use cases can be solved by Kubernetes alone, we do want to make them possible.
+
+### Deployment
+
+As a k8s operator, I want a to be able to add/remove new components (instances or types) to a new or running cluster (ideally without manually registering it with the apiserver).
+
+As a k8s operator, I want a kubectl command that allows me to validate that the cluster has been deployed and is fully functional.
+
+As a k8s operator, I want a kubectl command that allows me to check which components are deployed/installed on a specific cluster in order to infer which features it supports.
+
+Note: Inferring features from components isn't very useful with vanilla k8s, but becomes more useful as 3rd party or optional components are added (e.g. OpenShift) and/or if components-controller is split into multiple components.
+
+### Debugging
+
+As a k8s operator, I want a single endpoint (and kubectl command) where I can see the current condition of the cluster so that I can debug observed failures (dropped connections, unresponsiveness, etc.).
+
+Note: The endpoint should contain cause where possible, but at least the current condition of the components.
+
+### Notification
+
+As a k8s operator, I want to be notified when the cluster requires human intervention so that I can intervene.
+
+As an external notification system (e.g. PagerDuty), I want an endpoint that responds to regular requests with the current cluster condition so that I can determine (with my own rule set) when to notify the k8s operator.
```

Comment:
The current proposal doesn't quite replicate the `componentstatuses` API, but gets us pretty close.  With the latest proposal it is easy to identify the set of services that are "components" and make sure that some of the instances behind those services are ready.